### DISCUSSION TITLE
Startup time

### DISCUSSION THREAD

**Discussion by noahsmartin** [MobileNativeFoundation/discussions#37] [General]
Curious about how startup time is being measured (including pre-main) in production or as integration tests. Any preference to when startup time is considered finished, after the first frame is drawn?

**Comment by noahsmartin** [comment]
Here are a few things I've learned from measuring startup time:

1) In production consider the start point to be process start, not when applicationDidFinishLaunching or even main is called.

2) Third party frameworks dispatch to the main queue asynchronously during startup, it still blocks the main queue but you can't measure it if you consider the end time to be the end of applicationDidFinishLaunching. Instead I enqueue a block to run on the main queue as the last step of applicationDidFinishLaunching and mark its execution time as the end of app launch. This should account for any main queue work that third party libraries kick off asynchronously, as well as the time to render the first frame.

3) In development use DYLD_PRINT_STATISTICS to get a better idea of what causes pre-main time, but I haven't found a good way to tell at a detailed level what contributes to pre-main time in prod.

4) A more stable proxy for pre-main time in development can be the number of rebase operations dyld performs. This can be counted and attributed to each module, I have more detail on that [here](https://medium.com/codestory/why-swift-reference-types-are-bad-for-app-startup-time-90fbb25237fc).

I haven't heard of any good ways to track post-main startup time as part of the development process, but interested to hear what other people have!

**Reply by rudro** [reply]
We track pre-main startup time in production by timing the unix process start time to the earliest line of our code we can instrument. 

Currently, post-main is from the end of pre-main until a carefully placed stop point that needs to be maintained, i.e. something like map-visible for our Rider app or main-feed-visible for the Eats app. Definitely well past first frame.

We are also exploring a more dynamic post-main measurement which requires the main thread to be idle for a specified duration in order to consider startup complete. This would mean developers no longer need to carefully maintain the "stop point".

Regarding 3. we do see a wide range of pre-main times in productions (for the same hardware, OS, etc.) but have no way of breaking it down. We typically assume that certain things may be warm in dyld's caches etc. causing some low pre-main times.

**Reply by noahsmartin** [reply]
I've thought about an idle time heuristic as well, maybe using CFRunLoop observers. Was worried this could break if the user starts interacting with the app very quickly, or main thread bound animations play during app launch. Will be great to hear more if you get that working!

**Reply by keith** [reply]
For reference for folks here's a good post about measuring the pre-main time https://eisel.me/startup (we do this as well, but don't track them super closely)

**Reply by aaroncrespo** [reply]
The app I'm currently working on also measure the different phases: starting with `ProcessInfo` start, then `main.swift`, and then into the `didLaunch`/`willLaunch` events. We also try to capture main Thread Idle.

ProcessInfo to main usually gives us a rough idea of the impact of dependencies and dynamic loading on startup.

The numbers are inline with what we see from DYLD_PRINT_STATISTICS  but not as detailed, however can be measured in the wild. It also fits with Xcode's Launch Perf XCMetrics.

Maybe we should do a better job of determining when first frame is but since we track didLaunch and willLaunch and mainIdle. We should only be a few ticks off which should be within error%.

Data cleanliness is a challenge. For example we know cold starts are worse than `dyld3` cached  starts but we haven't looked at if its possible to segment them.

**Reply by uber-mahyar2** [reply]
>I've thought about an idle time heuristic as well, maybe using CFRunLoop observers. Was worried this could break if the user starts interacting with the app very quickly, or main thread bound animations play during app launch. Will be great to hear more if you get that working!

It works by recursively putting a block on the main dispatch queue and measuring the execution times between them.  Once 10 blocks have gaps under 20ms (200ms total 'idle' time) it stops and calls a callback.  We say startup completes in that callback.  So far it hasn't had any problems, but you have to be picky about when you start watching the main thread.   We also have a RIB tree (think view controller tree) and I attach the RIB tree to the startup span log.  This lets us see multiple kinds of 'startups' and catch the specific ones that have a longer-than-usual amount of time, which you can get from deep links and such.   If you do something similar, make sure to stable sort the view tree children of each node otherwise equivalent trees will look like different json blobs.

When you watch the main thread, you have to make sure you start watching when your startup procedure is mostly complete.  If you have a network wait inside your startup procedure, like our eats app, then you should only start watching the main thread after the idle network wait is complete.  Otherwise it's been fine.

**Comment by pilot34** [comment]
I was trying to make our local iOS measurements similar to Xcode Organizer numbers. Measuring time between process start time and first call of `viewDidLayoutSubviews` in the first visible `ViewController` gave the best results so far. My assumption was that `XCTApplicationLaunchMetric` in performance Xcode tests works similar to Xcode Organizer measurements. So I [checked](https://pilot34.medium.com/how-xcode-measures-app-launch-time-b5a66bd07c8c) what is included there. 

We also have another metric that shows time until user sees something meaningful on the screen, but that is a separate story.

**Reply by uber-mahyar2** [reply]
I have found any 'viewDid*' callback somewhat unreliable at times, like viewDidAppear.  Sometimes a view would visibly show up in my simulator and it would take half a second to seconds to actually call that code block.  I wouldn't necessarily rely on it unless you test it well.

**Comment by mackoj** [comment]
Did someone use [MetricKit](https://nshipster.com/metrickit/) for measuring this ?

**Reply by aaroncrespo** [reply]
We use it along side our metrics as a double check. The events we've written outside MetricKit are more fine grained so that's what we base our metrics on.

**Comment by omarzl** [comment]
We don't have that precision as other comments but I share what we do.
We measure the launch in each pull request using a `XCTApplicationLaunchMetric` test, we get the time from the log and then we send it to a prometheus push gateway where we can graph the progress across time in Grafana and we set Slack alerts when the launch increases.
We focus on monitoring it during the development phase to prevent that some change reaches production, of course since our tests run in a simulator with x86_64 arch we consider the values as reference.

**Comment by fabioalmeida** [comment]
At Farfetch we've been monitoring the startup time from pre-main to post-main using different approaches. 

We're leveraging `DYLD_PRINT_STATISTICS` to get more details on what's consuming time on pre-main (since this is what usually is more difficult to improve when compared do the tasks that affect post-main).
For post-main we are using an internal tool (that is now open-source as [NetworkMonitor](https://github.com/Farfetch/network-monitor-ios)) to mark certain points in time during our app start process and to extract all network calls being done as well.

We have a CI lane that runs in one of our CI machines that has a physical device attached to it (for the sake of simplicity, we only use 1 device and we review on a quarter basis what is the most common device/OS version used by our customers) and we get all the metrics from it leveraging `ios-deploy` and getting the log from the actual device.

Our experience showed us that results can fluctuate a lot depending on some pre-conditions therefore we make a distinction between the following:
* Cold start: app installed and run by the first time on the device
* Warm start: app already installed and opened once

Between both, even with the same conditions, the results can be different from run to run, so we perform 10 runs from each type in order to make it easier to discard the outliers. 

This is helping us to validate on every release how we are affecting the startup performance and to act accordingly if something is impacting it.
Just to give you an idea on what have we done since we started collecting this information, as our codebase is quite big and we do have a lot of dependencies in our app (~100, mostly internal), one of the biggest quick-wins we had so far on the startup performance (specially on the cold start) was to start linking our libraries statically instead of dynamically.

Hope this helps. Any question just shout.

**Reply by noahsmartin** [reply]
This sounds like a great setup thanks for sharing! I also have seen statically linking libraries be a big help for startup time and app size.

I'm curious about the cold vs. warm start and how much variance there is in each run. To simulate multiple runs of a cold start are you uninstalling and reinstalling the app from the device or are you rebooting the device each time you test a cold start? 

Do you track the average of the 10 runs or some other aggregation? I know many companies track percentiles in production (p90/p95) so it might make sense to take the max or second slowest of the 10 runs.

**Reply by fabioalmeida** [reply]
Between each run on the pre-main times the variance is usually around 50ms, but it general it can go from 10 to 100ms. In cold start launches this isn't much significant, but when we're analysing warm start a small change can be much more significant since the total value is much smaller when compared to cold starts.

For post-main the variance can sometimes be a bit higher because we have more variables that can influence the times, like network conditions for instance. 

To simulate multiple runs of a cold start we are just stopping and uninstalling the app from the device. Uninstalling, installing and running the app again is enough to see major differences on the startup time when compared to just closing and opening the app again (warm start). 

We gather all data on the runs, create some graphics with those and get some summary stats like mean, min, max, but yes, on a daily basis when everything is under the expected thresholds we just use percentiles (p95) for comparison.

**Comment by contrecc** [comment]
If you really, really want to understand your startup time and you are ok using jailbroken hardware -- look into a utility called spindump.  This utility takes stackshots from a process at a given interval and then displays that data in various forms.  When you run it on real hardware it gives you a true and complete view of the work being done during your startup process.

The utility exists on your mac too, type `man spindump` to get an idea of what it can do for you.  The version shipping on iPhones is a little different, type `spindump -?` to see its help as there are no man pages on iOS.
  
A simple way to get started would be a command like this:

spindump {part of bundle name} 10 1 -onlyTarget -symbolicate -wait

That will wait for your given bundle to launch, so you tap to launch it like a normal user, spindump will wait for you to do that.  Then it will take samples from all threads in the process at 1ms intervals for 10 seconds.  It will also symbolicate anything it can from the system libraries for you.

Happy profiling!

**Reply by ferologics** [reply]
Very cool, thanks for this hidden gem 💎 💯 🙏

**Reply by sumit-nathany** [reply]
This truly is a gem

**Comment by pyricau** [comment]
The conversation has been focused on iOS so far, but there's some good generic concepts that probably works for both, so I'll touch up on what we do at Square on Android for cold start (process start).

* Start: there are various tricks to obtain process fork time, which happens soon after the OS received the intent to start the app. Unfortunately some manufacturers enable a pool of pre forked idle processes. Process.getStartUptimeMillis() returns the time right before the APK is loaded, so it's not a good measure of "when the customer wanted to start the app" but it's a good measure of "the piont at which your app and its dependencies have any impact". Unfortunately sometimes it's also wildly off.
* End: the first frame rendering is definitely a good end time, unless the app has a dynamic loading nature (we don't) in which case you  probably want "the first frame after what we deem is a good start state". Unfortunately, there's no Android API to know the exact time at which the display showed a frame (or rather, the frame metrics APIs, give you no guarantee they'll call you on the next frame). My hack has been to hook into the next "draw" and from there post at the front of the main thread queue and capture  time there, which is a good enough approximation.

I wrote a series of deep dives on all these steps: https://dev.to/pyricau/series/7827

**Comment by tinder-owenthomas** [comment]
Good afternoon all! I'm curious if anyone has run into issues with pre-main startup time measurement now that iOS 15 pre-warms apps under certain conditions? Has anyone found a way to identify the time a member actually tapped the app on the home screen (or otherwise intentionally launched the app) for pre-warmed cases? And is there a good way to identify a pre-warm launch from a more traditional cold launch?

I know we can use MetricKit to get some general stats, but (having not tried it yet...) I don't think it gives granular (every launch) or app-specific (like time to completing a dispatched block from didFinishLaunching, or time to login screen load) data right?

**Reply by noahsmartin** [reply]
There's some discussion about this here: https://github.com/MobileNativeFoundation/discussions/discussions/146 I think the "ActivePrewarm" environment variable is what you're looking for, here's an example of using it: https://eisel.me/startup
